nselib/nbd.lua

```diff
@@ -0,0 +1,583 @@
+      stdnse.debug1("Failed to receive first 64 bits of magic from server: %s", rep)
+      self:close()
+      return false
+    end
+
+    if rep ~= NBD.magic.init_passwd then
+      stdnse.debug1("First 64 bits from server don't match expected magic: %s", stdnse.tohex(rep, {separator = ":"}))
+      self:close()
+      return false
+    end
+
+    local status, rep = self:receive(8)
+    if not status then
+      stdnse.debug1("Failed to receive second 64 bits of magic from server: %s", rep)
+      return false
+    end
+
+    if rep == NBD.magic.cliserv_magic_new then
+      self.protocol.negotiation = "newstyle"
+      return self:connect_new()
+    end
+
+    if rep == NBD.magic.cliserv_magic_old then
+      self.protocol.negotiation = "oldstyle"
+      return self:connect_old()
+    end
+
+    self.protocol.negotiation = "unrecognized"
+    stdnse.debug1("Second 64 bits from server don't match any known protocol magic: %s", stdnse.tohex(rep, {separator = ":"}))
+
+    self:close()
+
+    return true
+  end,
+
+  --- Cycles the connection to the server.
+  --
+  -- @name Comm.reconnect
+  --
+  -- @return status true on success, false on failure.
+  reconnect = function(self)
+    self:close()
+    return self:connect(self.connect_options)
+  end,
+
+  --- Attaches to an named share on the server.
+  --
+  -- @name Comm.attach
+  --
+  -- @return status true on success, false on failure.
+  attach = function(self, name)
+    assert(self.protocol.negotiation == "newstyle" or self.protocol.negotiation == "fixed newstyle")
+    assert(type(name) == "string")
+
+    local req = self:build_opt_req("EXPORT_NAME", {export_name = name})
+
+    local status, err = self:send(req)
+    if not status then
+      stdnse.debug1("Failed to send attach request for '%s': %s", name, err)
+      self:close()
+      return
+    end
+
+    local status, size = self:receive(8)
+    if not status then
+      stdnse.debug1("Failed to receive response to attach request for '%s': %s", name, size)
+      self:close()
+      return
+    end
+
+    local size, pos = (">I8"):unpack(size)
+    if pos ~= 9 then
+      stdnse.debug1("Failed to unpack size of exported block device from server.")
+      self:close()
+      return false
+    end
+
+    local status, tflags = self:receive(2)
+    if not status then
+      stdnse.debug1("Failed to receive transmission flags from server while attaching to export: %s", tflags)
+      self:close()
+      return false
+    end
+
+    local tflags, pos = (">I2"):unpack(tflags)
+    if pos ~= 3 then
+      stdnse.debug1("Failed to unpack transmission flags from server.")
+      self:close()
+      return false
+    end
+
+    tflags = self:parse_transmission_flags(tflags)
+
+    if self.protocol.zero_pad == "required" then
+      local status, err = self:receive(124)
+      if not status then
+	stdnse.debug1("Failed to receive zero pad from server while attaching to export: %s", err)
+	self:close()
+	return false
```

Comment:
Super-minor nitpick, but these 3 lines are tabs instead of spaces for indentation. A couple other places have this, too.